server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -779,10 +779,28 @@ public void relocated(
         final String targetAllocationId,
         final BiConsumer<ReplicationTracker.PrimaryContext, ActionListener<Void>> consumer,
         final ActionListener<Void> listener
+    ) throws IllegalIndexShardStateException, IllegalStateException {
+        relocated(targetNodeId, targetAllocationId, consumer, listener, null);
+    }
+
+    /**
+     * Provides an variant of {@link IndexShard#relocated(String, String, BiConsumer, ActionListener, Releasable)} with an option
+     * to relocate the shard under externally acquired primary permits.
+     *
+     * @param acquiredPrimaryPermits if null, waits until all the primary permits are acquired, otherwise it calls the consumer immediately
+     */
+    public void relocated(
+        final String targetNodeId,
+        final String targetAllocationId,
+        final BiConsumer<ReplicationTracker.PrimaryContext, ActionListener<Void>> consumer,
+        final ActionListener<Void> listener,
+        @Nullable final Releasable acquiredPrimaryPermits
```

Comment:
Just one question, are we planning on doing the error handling (i.e. not releasing the permits) in a follow-up PR?

Reply:
https://elasticco.atlassian.net/browse/ES-10573 should handle this

Reply:
Yes, that's the plan for a separate PR to acquire permits explicitly in order to pass them to swapped out `HollowIndexEngine`.